Refactoring Ex18 with preassembly option

dohyun-cse opened the pull request:
This PR is a sub-PR of #3598 with additional functionality, `preassembly` that linearize weak divergence for better performance in terms of computational time.

Preassembly turned on by default, and you can use `-mf` option to switch to nonlinear assembly.

dohyun-cse commented:
In my test (with previous wave speed for direct comparison)
- serial (`-r 4`): 160 sec (`master`), 158 sec (`-pa`), 300 sec (`-mf`)
- parallel (`-rs 2 -rp 2` with 8 threads):  24.8 sec (`master`), 23 sec (`-pa`), 42 sec (`-pa`).

Final errors with both `master` version and preassembly are the same, and the matrix-free gives is slightly more accurate than others. The updated version runs slightly faster possibly due to data locality and minor optimization in the flux function

v-dobrev commented:
@dohyun-cse, we use the following terminology when talking about various ways to apply an operator: https://github.com/mfem/mfem/blob/9a21f59e60b20d8a0cdde2af8d55be75dbc27bb2/fem/bilinearform.hpp#L28-L53

So if the divergence is pre-assembled as a `SparseMatrix` or `HypreParMatrix`, we call that "full assembly"; if only the element matrices are pre-assembled, we call that "element assembly", and when only data at quadrature points is pre-computed, we call that "partial assembly". In this context, using the term "pre-assemble" (and the option `-pa`) is a little under-specified and conflicts with the typical option we use for "partial assembly", `-pa` e.g. in `ex1`/`ex1p`. On the other hand, `-mf`, ("matrix-free") is exactly what we use for `AssemblyLevel::NONE`, i.e. "matrix-free", e.g. see `ex1`/`ex1p` in `miniapps/performance`.

With this terminology in in mind, it makes more sense to me to call the "pre-assembled" option "element assembly" (for the divergence operator) and use `-ea`/`--element-assembly-divergence` (or something like that) as the short/long command-line options. I believe the code is using "element assembly" and not "full assembly", right?

What do you think?

dohyun-cse commented:
> With this terminology in in mind, it makes more sense to me to call the "pre-assembled" option "element assembly" (for the divergence operator) and use `-ea`/`--element-assembly-divergence` (or something like that) as the short/long command-line options. I believe the code is using "element assembly" and not "full assembly", right?

Thank you for your suggestion. I also think that `-pa` is quite misleading as other examples use `pa` for partial assembly. I updated it and added some comment explaining the difference between two assembly procedures.

v-dobrev commented on examples/ex18.cpp:
```diff
@@ -95,9 +100,11 @@ int main(int argc, char *argv[])
+   args.AddOption(&preassembleWeakDiv, "-ea", "--element-assembly-divergence",
+                  "-mf", "--matrix-free-divergence",
+                  "Weak divergence assembly level\n"
+                  "    ea - Element assembly with linearized F\n"
```
I wouldn't call replacing the flux `F` with its interpolant `I_h F` "linearized F". "Interpolated F" is probably a better word to use.

v-dobrev commented on examples/ex18p.cpp:
```diff
@@ -103,9 +108,11 @@ int main(int argc, char *argv[])
+   args.AddOption(&preassembleWeakDiv, "-ea", "--element-assembly-divergence",
+                  "-mf", "--matrix-free-divergence",
+                  "Weak divergence assembly level\n"
+                  "    ea - Element assembly with linearized F\n"
```
Same remark here as in the serial version of the example: "linearized" -> "interpolated".

tarikdzanic commented on examples/ex18.hpp:
```diff
@@ -100,55 +108,119 @@ DGHyperbolicConservationLaws::DGHyperbolicConservationLaws(
-   DenseMatrix zmat, ymat;  // local dual vector storage
-
-   for (int i = 0; i < vfes.GetNE(); i++)
+   if (!weakdiv.empty()) // if weak divergence is pre-assembled
```
I think this comment is meant to say "if weak divergence is _not_ pre-assembled"

v-dobrev replied:
Actually, I think the comment was correct before: "weakdiv is not empty" <-> "weakdiv is precomputed".

dohyun-cse replied:
Oh, my bad. I didn't see that there is `!` when I revise the code. I revert it back the comment and added more comments on auxiliary variables. Please check 5699100631c3e6043263c4cdacc0e5d3b26116fb

tarikdzanic reviewed (approved):
LGTM besides a minor comment fix

cjvogl commented on examples/ex18.hpp:
```diff
@@ -100,55 +108,119 @@ DGHyperbolicConservationLaws::DGHyperbolicConservationLaws(
+      Vector xval, zval;
+      for (int i=0; i<vfes.GetNE(); i++)
+      {
+         auto Tr = vfes.GetElementTransformation(i);
```
```suggestion
         ElementTransformation* Tr = vfes.GetElementTransformation(i);
```

cjvogl reviewed (approved):
The looks good... I would avoid using `auto` in an example code (b/c example codes are effectively the user guide in MFEM), so I added a suggestion.
